activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -86,12 +86,21 @@ def build_relation(klass, attribute, value) # :nodoc:
       def scope_relation(record, relation)
         Array(options[:scope]).each do |scope_item|
           if reflection = record.class._reflect_on_association(scope_item)
```

Comment:
Good catch @matthewd! Just one [exception](https://github.com/rails/rails/pull/7072). But it looks much better after your suggestion.